Fix loc/iloc assignments when columns are selected

vnlitvinov opened the pull request:
<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/CONTRIBUTING.html
if you have questions about contributing.
-->

## What do these changes do?
The problem in current code is that it treats index and columns uniformely, while Pandas seems to be treating them differently.
Current code provides an indexing in the columns, but then checks for column names in the object being assigned (also `pandas.reindex()` expects column names, not column positions).

So I'm converting positions to names before trying to assign.

<!-- Please give a short brief about these changes. -->

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #1620
- [x] tests added and passing

dchigarev commented on modin/pandas/indexing.py:
```diff
@@ -391,12 +391,16 @@ def _broadcast_item(self, row_lookup, col_lookup, item, to_shape):
-                if not all(idx in item.columns for idx in col_lookup):
+                own_cols = self.qc.columns
+                # convert indices in column lookup to column names, as Pandas reindex expects them so
+                lookup = [own_cols[idx] for idx in col_lookup]
```
isn't `lookup = self.qc.columns[col_lookup]` will give us the same result?

vnlitvinov replied:
Indeed, I missed this point, thx

dchigarev commented on modin/pandas/indexing.py:
```diff
@@ -391,12 +391,16 @@ def _broadcast_item(self, row_lookup, col_lookup, item, to_shape):
                 )
             if hasattr(item, "columns"):
-                if not all(idx in item.columns for idx in col_lookup):
+                own_cols = self.qc.columns
+                # convert indices in column lookup to column names, as Pandas reindex expects them so
+                lookup = [own_cols[idx] for idx in col_lookup]
```
we probably should do the same changes for indices, because reindex expects index labels rather than its numeric indices, the same as it excepts the column labels.

New added test passes because in that particular case numeric indices equal to index labels, but if we'll try something like this it will crash:
```python
import modin.pandas as pd
import numpy as np

columns = ["col1", "col2", "col3"]
index = ["a", "b", "c", "d"]
df = pd.DataFrame({col: np.arange(4) for col in columns}, index=index)

df.loc[["a", "b"], :] = df.loc[["a", "b"], :].copy() # ValueError
```

vnlitvinov replied:
Nice, added to the test. Now it fails 😆

dchigarev commented:
It seems that reproducer from #1620 is still broken on that branch:
<details><summary> Spoiler </summary>


```python
import modin.pandas as pd
import pandas
import numpy as np
from modin.pandas.test.utils import df_equals, create_test_dfs

data = {
    "A": [np.nan, "c", "b", "a"],
    "B": ["d", "b", "a", "d"],
    "C": [1, 2, 3, 4],
    "D": [3, 4, 1, 3],
    "E": [1, 1, 1, 1],
}
md_df, pd_df = create_test_dfs(data)

md_df.loc[:, ["D", "C"]] = md_df.loc[:, ["D", "C"]].astype("float")
pd_df.loc[:, ["D", "C"]] = pd_df.loc[:, ["D", "C"]].astype("float")

df_equals(md_df, pd_df) # AssertionError
```

Output:
```
AssertionError: DataFrame.iloc[:, 2] (column name="C") are different

DataFrame.iloc[:, 2] (column name="C") values are different (100.0 %)
[index]: [0, 1, 2, 3]
[left]:  [3.0, 4.0, 1.0, 3.0]
[right]: [1.0, 2.0, 3.0, 4.0]
```


</details>


Error happens in cases when we're assigning columns in an order which differs from the order of columns in the frame (frame stores columns like `["C", "D"]` and we're trying to assign `["D", "C"]`).

It happens, because pandas backend internally sorts row and column lookups to determine to which partition every lookup index belong:
[`write_items`](https://github.com/modin-project/modin/blob/1a8cd0a3765df4cd9f76480e9862f522818333b2/modin/backends/pandas/query_compiler.py#L2860) -> [`_apply_select_indices`](https://github.com/modin-project/modin/blob/1a8cd0a3765df4cd9f76480e9862f522818333b2/modin/engines/base/frame/data.py#L1411) -> [`_get_dict_of_block_index`](https://github.com/modin-project/modin/blob/1a8cd0a3765df4cd9f76480e9862f522818333b2/modin/engines/base/frame/data.py#L846) -> [`sort`](https://github.com/modin-project/modin/blob/1a8cd0a3765df4cd9f76480e9862f522818333b2/modin/engines/base/frame/data.py#L919)

Since we're reindexing based on non-sorted lookups, we're getting a mismatch here.

vnlitvinov commented:
@dchigarev nice find with ordering, added to test. Now it fails xD

vnlitvinov commented:
UPD: I'm working on removing the need for index sorting.

vnlitvinov commented:
I've fixed most identified issues except the last one, which seem to require some redesign of internals.
I've extracted it into a separate GH tracker https://github.com/modin-project/modin/issues/2552 and `xfail`-ed the test that depends on it.

dchigarev commented on modin/pandas/test/dataframe/test_indexing.py:
```diff
@@ -411,6 +411,48 @@ def test_loc_assignment(index, columns):
+    select[axis] = sorted(pd_df.axes[axis][:-1], reverse=reverse_order)
+    select = tuple(select)
+
+    pd_df.loc[select] = pd_df.loc[select].copy()
+    md_df.loc[select] = md_df.loc[select].copy()
```
it seems that if 'loc' assignment doesn't work we will not see it in this test,
maybe we should do some modification for the right value to be sure that assignment is done? like:
```python
pd_df.loc[select] = pd_df.loc[select] + pd_df.loc[select]
md_df.loc[select] = md_df.loc[select] + md_df.loc[select]
```

vnlitvinov replied:
Makes sense, added

dchigarev reviewed (approved):
LGTM
